modules/Core/includes/endpoints/ListUsersEndpoint.php

```diff
@@ -15,71 +15,66 @@ public function __construct() {
-        } else {
-            $query = 'SELECT u.id, u.username, u.uuid, u.isbanned AS banned, u.active FROM nl2_users u';
-        }
-
         $operator = isset($_GET['operator']) && $_GET['operator'] == 'OR'
-            ? ' OR'
-            : ' AND';
+            ? ' OR '
+            : ' AND ';
 
         if (isset($_GET['group_id'])) {
-            $query .= ' INNER JOIN nl2_users_groups ug ON u.id = ug.user_id WHERE ug.group_id = ?';
+            $query .= ' INNER JOIN nl2_users_groups ug ON u.id = ug.user_id';
+            $where[] = 'ug.group_id = ?';
             $params[] = $_GET['group_id'];
-            $filterGroup = true;
+        }
+
+        if (isset($_GET['integration'])) {
+            $query .= ' INNER JOIN nl2_users_integrations ui ON ui.user_id=u.id INNER JOIN nl2_integrations i ON i.id=ui.integration_id';
+            $where[] = 'i.name = ?';
+            $params[] = $_GET['integration'];
         }
 
         if (isset($_GET['banned'])) {
-            if (isset($filterGroup)) {
-                $query .= $operator;
-            } else {
-                $query .= ' WHERE';
-            }
-            $query .= '`u.isbanned` = ' . ($_GET['banned'] == 'true' ? '1' : '0');
-            $filterBanned = true;
+            $where[] = '`u`.`isbanned` = ' . ($_GET['banned'] == 'true' ? '1' : '0');
         }
 
         if (isset($_GET['active'])) {
-            if (isset($filterBanned) || isset($filterGroup)) {
-                $query .= $operator;
-            } else {
-                $query .= ' WHERE';
-            }
-            $query .= ' `u.active` = ' . ($_GET['active'] == 'true' ? '1' : '0');
-            $filterActive = true;
+            $where[] = '`u`.`active` = ' . ($_GET['active'] == 'true' ? '1' : '0');
         }
 
-        if ($discord_enabled && isset($_GET['discord_linked'])) {
-            if (isset($filterBanned) || isset($filterActive) || isset($filterGroup)) {
-                $query .= $operator;
-            } else {
-                $query .= ' WHERE';
-            }
-            $query .= ' `u.discord_id` IS ' . ($_GET['discord_linked'] == 'true' ? 'NOT' : '') . ' NULL';
+        // Build where string
+        $where_string = ' WHERE ';
+        foreach ($where as $item) {
+            $where_string .= $item . $operator;
         }
+        $where_string = rtrim($where_string, $operator);
 
-        $users = $api->getDb()->selectQuery($query, $params)->results();
+        $users = $api->getDb()->selectQuery($query . $where_string, $params)->results();
 
         $users_json = [];
         foreach ($users as $user) {
+            $integrations = [];
+            $integrations_query = $api->getDb()->selectQuery('SELECT ui.*, i.name FROM nl2_users_integrations ui INNER JOIN nl2_integrations i ON i.id=ui.integration_id WHERE user_id = ? AND username IS NOT NULL AND identifier IS NOT NULL', [$user->id])->results();
```

Comment:
this is an O^n situation (1 more query for each user) - is it not possible to include this in the main query above?

Reply:
https://github.com/NamelessMC/Nameless/issues/2683